How many late days are you using for this assignment? 

0 days

In [3]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [4]:
# Load the given datasets
# Print the data
numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")

print("numerical CSV File:")
print(numerical)

print("\categorical CSV File:")
print(categorical)


numerical CSV File:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 
For numerical chart:
BP (Blood Pressure)
Critical for managing CKD progression
BGR (Blood Urea Nitrogen to Glucose Ratio)
A predictor of 1-year mortality risk in CKD patients. 
BU (Blood Urea Nitrogen)
Reflects kidney filtration capacity. 
SC (Serum Creatinine)
A primary marker for estimating glomerular filtration rate (eGFR), which determines CKD staging 
Sod (Sodium)
Imbalances (hyponatremia or hypernatremia) are common in CKD due to reduced kidney capacity to regulate fluid and electrolytes.
Pot (Potassium)
Hyperkalemia (high potassium) is a major concern in advanced CKD,
Hemo (Hemoglobin)
Anemia is common in CKD due to reduced erythropoietin production. 
PCV (Packed Cell Volume / Hematocrit)
Reflects red blood cell mass.
WBCC (White Blood Cell Count)
Elevated levels may indicate infection or inflammation
RBCC (Red Blood Cell Count)
Reduced counts are linked to CKD-associated anemia. 

For categorical chart:
AL (Albuminuria)
SU (Serum Urate/Uric Acid)
RBC (Red Blood Cell Count)
PC (Pus Cells)
PCC (Pus Cell Clumps)
BA (Bacteria)
HTN (Hypertension)
DM (Diabetes Mellitus)
CAD (Coronary Artery Disease)
Appet (Appetite)
PE (Pulmonary Embolism)
ANE (Anemia)
Target (Clinical Outcome)

Shared features:
Only RBC is shared in both chart. 
But both charts have some overlaps features:
Blood pressure control (HTN, BP).
Anemia treatment (Hemo, RBC, ANE).
Infection management (PC, BA).



# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [5]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates = numerical.duplicated(subset=['unique_id'])
num_duplicates_numerical = duplicates.sum()
print(f"Number of duplicate rows in numerical dataset: {num_duplicates_numerical}")
# Drop these duplicate rows
print(f"Number of rows of numerical chart before dropping duplicates: {len(numerical)}")
numerical_cleaned = numerical.drop_duplicates(subset=['unique_id'])
print(f"Number of rows of numerical chart after dropping duplicates: {len(numerical_cleaned)}")

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates = categorical.duplicated(subset=['unique_id'])
num_duplicates_categorical = duplicates.sum()
print(f"Number of duplicate rows in categorical dataset: {num_duplicates_categorical}")
# Drop these duplicate rows
print(f"Number of rows of categorical chart before dropping duplicates: {len(categorical)}")
categorical_cleaned = categorical.drop_duplicates(subset=['unique_id'])
print(f"Number of rows of categorical chart after dropping duplicates: {len(categorical_cleaned)}")



Number of duplicate rows in numerical dataset: 30
Number of rows of numerical chart before dropping duplicates: 420
Number of rows of numerical chart after dropping duplicates: 390
Number of duplicate rows in categorical dataset: 30
Number of rows of categorical chart before dropping duplicates: 430
Number of rows of categorical chart after dropping duplicates: 400


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, since two datasets have different number of rows, so there will be ids not present in both datasets. The rows are missing from one would have NaN in features they do not have.

In [6]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_chart = pd.merge(numerical, categorical, on='unique_id', how='outer')

# 'inner' – keeps only rows with matching 'unique_id' in both DataFrames
# 'outer' – keeps all rows from both DataFrames, fills with NaN where there are no matches
# 'left' – keeps all rows from df1, including all rows from df2 that match
# 'right' – keeps all rows from df2, including all rows from df1 that match

#Print the combined dataset
print(merged_chart)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
487     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
488     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
489     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
490     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
491     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [7]:
# Calculate the percentage of rows that contain atleast one missing value

rows_with_nan = merged_chart.isna().any(axis=1)  # Create a boolean Series
num_rows_with_nan = rows_with_nan.sum()  # Count True values

print(f"\nNumber of rows containing at least one NaN: {num_rows_with_nan}")
num_rows = len(merged_chart)

# Print %
percent_NaN = 100*(num_rows_with_nan)/num_rows
print(f"\nPercentage of rows with missing values: {percent_NaN}%")

# Drop these rows from the dataset
merged_chart_cleaned = merged_chart.dropna()


# Print the Dataset
print("\nCleaned DataFrame (without NaN):")
print(merged_chart_cleaned)



Number of rows containing at least one NaN: 294

Percentage of rows with missing values: 59.75609756097561%

Cleaned DataFrame (without NaN):
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
482     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
483     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
486     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
487     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  2

## Part 3.4: Sort the dataset according to the Labels

In [8]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged_chart_cleaned_sorted = merged_chart_cleaned.sort_values(by='Target')

# Step 2: Reset the index after sorting
merged_chart_cleaned_sorted.reset_index(drop=True, inplace=True)

# Print the dataset
print("\nSorted DataFrame (by 'Target' column):")
print(merged_chart_cleaned_sorted)



Sorted DataFrame (by 'Target' column):
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       234299  48.0  80.0   95.0  163.0   7.7  136.0  3.8   9.8  32.0  ...   
2       636769  50.0  90.0  128.0  208.0   9.2  134.0  4.8   8.2  22.0  ...   
3       484175  53.0  90.0   70.0  107.0   7.2  114.0  3.7   9.5  29.0  ...   
4       481293  60.0  60.0  288.0   36.0   1.7  130.0  3.0   7.9  25.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
193     433613  50.0  80.0   97.0   40.0   0.6  150.0  4.5  14.2  48.0  ...   
194     436056  12.0  80.0  100.0   26.0   0.6  137.0  4.4  15.8  49.0  ...   
195     441170  79.0  80.0  111.0   44.0   1.2  146.0  3.6  16.3  40.0  ...   
196     441170  79.0  80.0  111.0   44.0   1.2  146.0  3.6  16.3  40.0  ...   
197     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [9]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

merged_chart_cleaned_sorted_with_dummies = pd.get_dummies(merged_chart_cleaned_sorted, columns=categorical_columns)
# If the dummy values are still boolean, ensuring conversion to integers
merged_chart_cleaned_sorted_with_dummies = merged_chart_cleaned_sorted_with_dummies.astype(int)

# Print the dataset
print(merged_chart_cleaned_sorted_with_dummies)

     unique_id  age  bp  bgr   bu  sc  sod  pot  hemo  pcv  ...  cad_no  \
0       102677   59  70   76  186  15  135    7     7   22  ...       1   
1       234299   48  80   95  163   7  136    3     9   32  ...       1   
2       636769   50  90  128  208   9  134    4     8   22  ...       1   
3       484175   53  90   70  107   7  114    3     9   29  ...       1   
4       481293   60  60  288   36   1  130    3     7   25  ...       1   
..         ...  ...  ..  ...  ...  ..  ...  ...   ...  ...  ...     ...   
193     433613   50  80   97   40   0  150    4    14   48  ...       1   
194     436056   12  80  100   26   0  137    4    15   49  ...       1   
195     441170   79  80  111   44   1  146    3    16   40  ...       1   
196     441170   79  80  111   44   1  146    3    16   40  ...       1   
197     996076   57  60  105   49   1  150    4    15   44  ...       1   

     cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0          0           0   

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: If a categorical variable can take on more than two values, it means this row can be measured more accurately based on this variable. Additionally, this indicates that this variable will provide more information than just a simple yes or no.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [10]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

merged_predropped = merged_chart_cleaned_sorted_with_dummies

mean = merged_predropped[numerical_columns].mean()
std_dev = merged_predropped[numerical_columns].std()

# Define the upper and lower bounds for outlier detection
upper_limit = mean + 3 * std_dev
lower_limit = mean - 3 * std_dev

# Filter out rows with outliers
# Create a boolean condition for retaining rows within the limits
condition = (merged_predropped[numerical_columns] <= upper_limit) & (merged_predropped[numerical_columns] >= lower_limit)

# Use .all(axis=1) to ensure all specified columns meet the condition for each row
merged_dropped_outlier = merged_predropped[condition.all(axis=1)]
merged_dropped_outlier = merged_dropped_outlier.reset_index()
# Print the dataset
print(merged_dropped_outlier)


     index  unique_id  age  bp  bgr   bu  sc  sod  pot  hemo  ...  cad_no  \
0        1     234299   48  80   95  163   7  136    3     9  ...       1   
1        4     481293   60  60  288   36   1  130    3     7  ...       1   
2        6     474407   21  90  107   40   1  125    3     8  ...       1   
3        7     843362   71  60  118  125   5  136    4    11  ...       1   
4        8     828592   71  70  219   82   3  133    4    10  ...       0   
..     ...        ...  ...  ..  ...  ...  ..  ...  ...   ...  ...     ...   
168    193     433613   50  80   97   40   0  150    4    14  ...       1   
169    194     436056   12  80  100   26   0  137    4    15  ...       1   
170    195     441170   79  80  111   44   1  146    3    16  ...       1   
171    196     441170   79  80  111   44   1  146    3    16  ...       1   
172    197     996076   57  60  105   49   1  150    4    15  ...       1   

     cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0  

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [11]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
scaler = MinMaxScaler()
normalized = merged_dropped_outlier
normalized[numerical_columns] = scaler.fit_transform(merged_dropped_outlier[numerical_columns])

# Print the dataset
print(normalized)



     index  unique_id       age    bp       bgr        bu     sc       sod  \
0        1     234299  0.545455  0.50  0.107296  1.000000  0.875  0.600000   
1        4     481293  0.701299  0.00  0.935622  0.169935  0.125  0.428571   
2        6     474407  0.194805  0.75  0.158798  0.196078  0.125  0.285714   
3        7     843362  0.844156  0.00  0.206009  0.751634  0.625  0.600000   
4        8     828592  0.844156  0.25  0.639485  0.470588  0.375  0.514286   
..     ...        ...       ...   ...       ...       ...    ...       ...   
168    193     433613  0.571429  0.50  0.115880  0.196078  0.000  1.000000   
169    194     436056  0.077922  0.50  0.128755  0.104575  0.000  0.628571   
170    195     441170  0.948052  0.50  0.175966  0.222222  0.125  0.885714   
171    196     441170  0.948052  0.50  0.175966  0.222222  0.125  0.885714   
172    197     996076  0.662338  0.00  0.150215  0.254902  0.125  1.000000   

          pot  hemo  ...  cad_no  cad_yes  appet_good  appet_po

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [12]:
# Remove that column
# The index is not needed since there is indice. Other than that, every redundant binary feature can be removed since 
# these are mutually exclusive categories derived from single variables.

columns_to_drop = ['index','rbc_abnormal', 'pc_abnormal', 'pcc_notpresent', 'ba_notpresent', 'htn_no', 'dm_no', 'cad_no', 'appet_poor', 'pe_no', 'ane_no', 'Target_notckd']
final_chart = normalized
final_chart.drop(columns=columns_to_drop,inplace=True)

# Print the dataset

print(final_chart)

     unique_id       age    bp       bgr        bu     sc       sod       pot  \
0       234299  0.545455  0.50  0.107296  1.000000  0.875  0.600000  0.333333   
1       481293  0.701299  0.00  0.935622  0.169935  0.125  0.428571  0.333333   
2       474407  0.194805  0.75  0.158798  0.196078  0.125  0.285714  0.333333   
3       843362  0.844156  0.00  0.206009  0.751634  0.625  0.600000  0.666667   
4       828592  0.844156  0.25  0.639485  0.470588  0.375  0.514286  0.666667   
..         ...       ...   ...       ...       ...    ...       ...       ...   
168     433613  0.571429  0.50  0.115880  0.196078  0.000  1.000000  0.666667   
169     436056  0.077922  0.50  0.128755  0.104575  0.000  0.628571  0.666667   
170     441170  0.948052  0.50  0.175966  0.222222  0.125  0.885714  0.333333   
171     441170  0.948052  0.50  0.175966  0.222222  0.125  0.885714  0.333333   
172     996076  0.662338  0.00  0.150215  0.254902  0.125  1.000000  0.666667   

     hemo       pcv  ...  p

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [13]:
# Export the dataframe to a new csv file
csv_file_path = 'data.csv'  # Specify the path for your CSV
final_chart.to_csv(csv_file_path, index=False)  # Set index=False to not include the index in the CSV
print(f"DataFrame exported to {csv_file_path}.")

# Export the dataframe to a new json file
json_file_path = 'data.json'  # Specify the path for your JSON
final_chart.to_json(json_file_path, orient='records', lines=False)  # Change orient and lines as necessary
print(f"DataFrame exported to {json_file_path}.")

DataFrame exported to data.csv.
DataFrame exported to data.json.


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [14]:
print(final_chart.head(15))

    unique_id       age    bp       bgr        bu     sc       sod       pot  \
0      234299  0.545455  0.50  0.107296  1.000000  0.875  0.600000  0.333333   
1      481293  0.701299  0.00  0.935622  0.169935  0.125  0.428571  0.333333   
2      474407  0.194805  0.75  0.158798  0.196078  0.125  0.285714  0.333333   
3      843362  0.844156  0.00  0.206009  0.751634  0.625  0.600000  0.666667   
4      828592  0.844156  0.25  0.639485  0.470588  0.375  0.514286  0.666667   
5      343710  0.714286  0.50  0.442060  0.901961  0.375  0.571429  1.000000   
6      438182  0.870130  0.50  0.785408  0.862745  0.500  0.657143  1.000000   
7      397388  0.740260  1.00  0.901288  0.163399  0.375  0.800000  0.333333   
8      899916  1.000000  0.25  0.137339  0.326797  0.250  0.000000  1.000000   
9      995177  0.818182  0.25  0.618026  0.562092  0.750  0.142857  0.333333   
10     475200  0.753247  1.00  0.399142  0.287582  0.875  0.714286  0.666667   
11     137148  0.870130  1.00  0.965665 

Paste here: 
### Paste the markdown table here
  unique_id       age    bp       bgr        bu     sc       sod       pot  \
0      234299  0.545455  0.50  0.107296  1.000000  0.875  0.600000  0.333333   
1      481293  0.701299  0.00  0.935622  0.169935  0.125  0.428571  0.333333   
2      474407  0.194805  0.75  0.158798  0.196078  0.125  0.285714  0.333333   
3      843362  0.844156  0.00  0.206009  0.751634  0.625  0.600000  0.666667   
4      828592  0.844156  0.25  0.639485  0.470588  0.375  0.514286  0.666667   
5      343710  0.714286  0.50  0.442060  0.901961  0.375  0.571429  1.000000   
6      438182  0.870130  0.50  0.785408  0.862745  0.500  0.657143  1.000000   
7      397388  0.740260  1.00  0.901288  0.163399  0.375  0.800000  0.333333   
8      899916  1.000000  0.25  0.137339  0.326797  0.250  0.000000  1.000000   
9      995177  0.818182  0.25  0.618026  0.562092  0.750  0.142857  0.333333   
10     475200  0.753247  1.00  0.399142  0.287582  0.875  0.714286  0.666667   
11     137148  0.870130  1.00  0.965665  0.522876  0.625  0.714286  0.000000   
12     129053  0.545455  0.50  0.270386  0.843137  1.000  0.485714  1.000000   
13     546225  0.519481  0.00  0.399142  0.535948  0.375  0.742857  0.666667   
14     621332  0.688312  0.50  1.000000  0.163399  0.125  0.200000  0.333333   

    hemo       pcv  ...  pc_normal  pcc_present  ba_present  htn_yes  dm_yes  \
0    0.2  0.290323  ...          0            0           0        1       0   
1    0.0  0.064516  ...          0            1           0        1       0   
2    0.1  0.000000  ...          0            1           1        0       0   
3    0.4  0.387097  ...          1            0           0        1       1   
4    0.3  0.322581  ...          0            1           1        1       1   
5    0.0  0.032258  ...          0            0           0        1       1   
6    0.3  0.322581  ...          0            0           0        1       1   
7    0.6  0.548387  ...          1            0           1        1       0   
8    0.1  0.096774  ...          1            0           0        1       0   
9    0.2  0.161290  ...          0            1           1        1       1   
10   0.0  0.096774  ...          0            0           1        1       1   
11   0.2  0.225806  ...          0            1           0        1       1   
12   0.3  0.419355  ...          0            0           1        0       1   
13   0.2  0.161290  ...          1            0           0        1       1   
14   0.3  0.387097  ...          1            0           0        0       1   

    cad_yes  appet_good  pe_yes  ane_yes  Target_ckd  
0         0           1       0        1           1  
1         0           0       0        1           1  
2         0           1       0        1           1  
3         0           0       1        0           1  
4         1           1       0        0           1  
5         1           0       1        1           1  
6         1           1       0        0           1  
7         1           1       0        0           1  
8         0           0       0        1           1  
9         1           1       1        1           1  
10        0           1       1        0           1  
11        1           0       0        0           1  
12        0           1       1        0           1  
13        0           1       0        0           1  
14        0           0       0        0           1  

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Instead of entering whole datasets, I only enetered the part of data that I want the LLM to notices. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [15]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'bgr': 'Blood Glucose Ratio',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_normal': 'RBC Normal',
    'pc_normal': 'PC Normal',
    'pcc_present': 'PCC Present',
    'ba_present': 'BA Present',
    'htn_yes': 'Hypertension Yes',
    'dm_yes': 'Diabetes Mellitus Yes',
    'cad_yes': 'Coronary Artery Disease Yes',
    'appet_good': 'Appetite Good',
    'pe_yes': 'PE Yes',
    'ane_yes': 'Anemia Yes',
}
final_chart.rename(columns=rename_mapping, inplace=True)

print(final_chart)

     unique_id       Age  Blood Pressure  Blood Glucose Ratio  Blood Urea  \
0       234299  0.545455            0.50             0.107296    1.000000   
1       481293  0.701299            0.00             0.935622    0.169935   
2       474407  0.194805            0.75             0.158798    0.196078   
3       843362  0.844156            0.00             0.206009    0.751634   
4       828592  0.844156            0.25             0.639485    0.470588   
..         ...       ...             ...                  ...         ...   
168     433613  0.571429            0.50             0.115880    0.196078   
169     436056  0.077922            0.50             0.128755    0.104575   
170     441170  0.948052            0.50             0.175966    0.222222   
171     441170  0.948052            0.50             0.175966    0.222222   
172     996076  0.662338            0.00             0.150215    0.254902   

     Serum Creatinine    Sodium  Potassium  Hemoglobin  Packed Cell Volume 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [16]:
# Converted SQL to Pandas code

# Sample DataFrame creation 
data = {
    'Target': ['ckd', 'notckd', 'ckd', 'notckd', 'ckd', 'notckd'],
}
df = pd.DataFrame(data)

# Grouping by the 'Target' column and counting occurrences
result = df.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)




   Target  count
0     ckd      3
1  notckd      3


*Describe what the above code does here*
Seems like this part of SQL server as a counter that counts how many times target appears. The overall flow of the python code mimics the SQL query. It groups the data based on unique values in the Target column and counts the occurrences of those values

